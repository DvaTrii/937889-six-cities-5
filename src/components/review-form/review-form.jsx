import React from "react";
import {withRouter} from "react-router-dom";
import {compose} from "redux";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {postReview} from "../../store/data/operations";
import {getAuthorizationStatus} from "../../store/user/selectors";
import {getOfferIdFromUrl} from "../../store/data/selectors";

class ReviewForm extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      rating: ``,
      review: ``,
      formIsValid: false,
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleFieldChange = this.handleFieldChange.bind(this);
  }

  validateForm() {
    const isValid = this.state.rating && (this.state.review.length < 300 && this.state.review.length > 50);

    this.setState({
      formIsValid: isValid
    });
  }

  handleSubmit(evt) {
    evt.preventDefault();
    this.props.postReview(this.state, this.props.id);
    this.setState({
      rating: ``,
      review: ``,
      formIsValid: false,
    });
  }

  handleFieldChange(evt) {
    const {name, value} = evt.target;
    this.setState({[name]: value}, this.validateForm);
  }

  render() {
    return (
      <form
        className="reviews__form form"
        action="#"
        method="post"
        onSubmit={this.handleSubmit}>
        <label className="reviews__label form__label" htmlFor="review">Your review</label>
        <div className="reviews__rating-form form__rating">
          <input
            className="form__rating-input visually-hidden"
            name="rating"
            value="5"
            id="5-stars"
            type="radio"
            onChange={this.handleFieldChange}
            checked={this.state.rating === `5`}
          />
          <label htmlFor="5-stars" className="reviews__rating-label form__rating-label" title="perfect">
            <svg className="form__star-image" width="37" height="33">
              <use xlinkHref="#icon-star"></use>
            </svg>
          </label>

          <input
            className="form__rating-input visually-hidden"
            name="rating"
            value="4"
            id="4-stars"
            type="radio"
            onChange={this.handleFieldChange}
            checked={this.state.rating === `4`}
          />
          <label htmlFor="4-stars" className="reviews__rating-label form__rating-label" title="good">
            <svg className="form__star-image" width="37" height="33">
              <use xlinkHref="#icon-star"></use>
            </svg>
          </label>

          <input
            className="form__rating-input visually-hidden"
            name="rating"
            value="3"
            id="3-stars"
            type="radio"
            onChange={this.handleFieldChange}
            checked={this.state.rating === `3`}
          />
          <label htmlFor="3-stars" className="reviews__rating-label form__rating-label" title="not bad">
            <svg className="form__star-image" width="37" height="33">
              <use xlinkHref="#icon-star"></use>
            </svg>
          </label>

          <input
            className="form__rating-input visually-hidden"
            name="rating"
            value="2"
            id="1-stars"
            type="radio"
            onChange={this.handleFieldChange}
            checked={this.state.rating === `2`}
          />
          <label htmlFor="2-stars" className="reviews__rating-label form__rating-label" title="badly">
            <svg className="form__star-image" width="37" height="33">
              <use xlinkHref="#icon-star"></use>
            </svg>
          </label>

          <input
            className="form__rating-input visually-hidden"
            name="rating"
            value="1"
            id="1-stars"
            type="radio"
            onChange={this.handleFieldChange}
            checked={this.state.rating === `1`}
          />
          <label htmlFor="1-star" className="reviews__rating-label form__rating-label" title="terribly">
            <svg className="form__star-image" width="37" height="33">
              <use xlinkHref="#icon-star"></use>
            </svg>
          </label>
        </div>
        <textarea
          className="reviews__textarea form__textarea"
          id="review"
          name="review"
          placeholder="Tell how was your stay, what you like and what can be improved"
          onChange={this.handleFieldChange}
          value={this.state.review}
        />
        <div className="reviews__button-wrapper">
          <p className="reviews__help">
            To submit review please make sure to set <span className="reviews__star">rating</span> and describe
            your stay with at least <b className="reviews__text-amount">50 characters</b>.
          </p>
          <button
            className="reviews__submit form__submit button"
            type="submit"
            disabled={!this.state.formIsValid}>
            Submit</button>
        </div>
      </form>

    );
  }
}

ReviewForm.propTypes = {
  postReview: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
  authorizationStatus: PropTypes.bool.isRequired,
};

const mapStateToProps = (state, ownProps) => {
  return ({
    id: getOfferIdFromUrl(ownProps),
    authorizationStatus: getAuthorizationStatus(state),
  });
};

const mapDispatchToProps = (dispatch) => ({
  postReview(data, id) {
    dispatch(postReview(data, id));
  }
});

export {ReviewForm};
export default compose(withRouter, connect(mapStateToProps, mapDispatchToProps))(ReviewForm);
