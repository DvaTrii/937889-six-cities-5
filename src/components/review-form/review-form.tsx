import * as React from "react";
import {withRouter} from "react-router-dom";
import {compose} from "redux";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {postReview} from "../../store/data/operations";
import {getOfferIdFromUrl} from "../../store/data/selectors";
import {extend} from "../../utils";
import {ReviewSymbolsNumber} from "../../const";

const ReviewForm = ({offerId, postCurrentReview}) => {

  const [review, setReview] = React.useState({
    rating: ``,
    review: ``
  });

  const [isValidForm, setIsValidForm] = React.useState(false);

  React.useEffect(() => {
    validateForm();
  }, [review]);

  const handleSubmit = (evt) => {
    evt.preventDefault();
    postCurrentReview(review, offerId);
    setReview(() => {
      return {
        rating: ``,
        review: ``
      };
    });
    setIsValidForm(false);
  };

  const validateForm = () => {
    const isValid = review.rating &&
      (review.review.length >= ReviewSymbolsNumber.MIN
        && review.review.length <= ReviewSymbolsNumber.MAX);
    setIsValidForm(isValid);
  };

  const handleFieldChange = (evt) => {
    const {name, value} = evt.target;
    setReview(extend(review, {[name]: value}));
  };

  return (
    <form
      className="reviews__form form"
      action="#"
      method="post"
      onSubmit={handleSubmit}>
      <label className="reviews__label form__label" htmlFor="review">Your review</label>
      <div className="reviews__rating-form form__rating">
        <input
          className="form__rating-input visually-hidden"
          name="rating"
          value="5"
          id="5-stars"
          type="radio"
          onChange={handleFieldChange}
          checked={review.rating === `5`}
        />
        <label htmlFor="5-stars" className="reviews__rating-label form__rating-label" title="perfect">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input
          className="form__rating-input visually-hidden"
          name="rating"
          value="4"
          id="4-stars"
          type="radio"
          onChange={handleFieldChange}
          checked={review.rating === `4`}
        />
        <label htmlFor="4-stars" className="reviews__rating-label form__rating-label" title="good">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input
          className="form__rating-input visually-hidden"
          name="rating"
          value="3"
          id="3-stars"
          type="radio"
          onChange={handleFieldChange}
          checked={review.rating === `3`}
        />
        <label htmlFor="3-stars" className="reviews__rating-label form__rating-label" title="not bad">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input
          className="form__rating-input visually-hidden"
          name="rating"
          value="2"
          id="2-stars"
          type="radio"
          onChange={handleFieldChange}
          checked={review.rating === `2`}
        />
        <label htmlFor="2-stars" className="reviews__rating-label form__rating-label" title="badly">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input
          className="form__rating-input visually-hidden"
          name="rating"
          value="1"
          id="1-stars"
          type="radio"
          onChange={handleFieldChange}
          checked={review.rating === `1`}
        />
        <label htmlFor="1-star" className="reviews__rating-label form__rating-label" title="terribly">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>
      </div>
      <textarea
        className="reviews__textarea form__textarea"
        id="review"
        name="review"
        placeholder="Tell how was your stay, what you like and what can be improved"
        onChange={handleFieldChange}
        value={review.review}
      />
      <div className="reviews__button-wrapper">
        <p className="reviews__help">
          To submit review please make sure to set <span className="reviews__star">rating</span> and describe
          your stay with at least <b className="reviews__text-amount">50 characters</b>.
        </p>
        <button
          className="reviews__submit form__submit button"
          type="submit"
          disabled={!isValidForm}>
          Submit</button>
      </div>
    </form>

  );
};

ReviewForm.propTypes = {
  postCurrentReview: PropTypes.func.isRequired,
  offerId: PropTypes.number.isRequired,
};

const mapStateToProps = (state, ownProps) => {
  return ({
    offerId: getOfferIdFromUrl(ownProps),
  });
};

const mapDispatchToProps = (dispatch) => ({
  postCurrentReview(data, id) {
    dispatch(postReview(data, id));
  }
});

export {ReviewForm};
export default compose(withRouter, connect(mapStateToProps, mapDispatchToProps))(ReviewForm);
