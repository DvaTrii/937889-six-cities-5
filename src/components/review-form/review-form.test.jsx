import React from "react";
import renderer from "react-test-renderer";
import {ReviewForm} from "./review-form";

it(`Should ReviewForm render properly`, () => {
  const offerId = 3;

  const tree = renderer
    .create(
        <ReviewForm
          postReview={() => {}}
          offerId={offerId}
        />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
