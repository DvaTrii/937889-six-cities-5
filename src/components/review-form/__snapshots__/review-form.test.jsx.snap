// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should ReviewForm render properly 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      id="5-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value="5"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="5-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      id="4-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value="4"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="4-stars"
      title="good"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      id="3-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value="3"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="3-stars"
      title="not bad"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      id="1-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value="2"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="2-stars"
      title="badly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      checked={false}
      className="form__rating-input visually-hidden"
      id="1-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value="1"
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="1-star"
      title="terribly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    id="review"
    name="review"
    onChange={[Function]}
    placeholder="Tell how was your stay, what you like and what can be improved"
    value=""
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set 
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least 
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled={true}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
